### PR TITLE
JDBC Sources can discover ints and longs + handle them in normalization

### LINKED ISSUE #12606: 🐛 JDBC source connectors: integer are converted to float on destination
## Environment

- **Source Connector and version**: All database JDBC connector

## Current Behavior
JDBC connectors are typing `integer` as `number` in the source catalog. `number` are casted to `float` / `decimal` type on destination.

## Expected Behavior
`integer` in source database schemas should be defined as `integer` in the catalog schema that is generated by the source connector and not `number`.

### PR DESCRIPTION
**Pull request by edgao**
This is a pretty hefty PR, so definitely LMK If you want to chat about it live.

`protocol-reviewers` is tagged solely for the [JsonSchemaType](https://github.com/airbytehq/airbyte/pull/14362/files#diff-96f14a3d9a899323518d3d7631dd318db81bad0a94aeaf35f39edb9b58252ebc) update - everything else is connectors-specific.

## What
Closes https://github.com/airbytehq/airbyte/issues/12606, https://github.com/airbytehq/airbyte/issues/12057

## How
* JDBC sources now distinguish between integral and floating-point types
* Normalization now correctly maps int and big_integer types.
    * normalization has a new `type_very_large_integer` type, to handle big_integers.

Note that S3 Avro/Parquet now treat big_integer as string-typed (previously int64). This aligns with our destination acceptance tests, as well as existing documentation.

Summarizing the new behavior:
* `{type: integer}` is treated by normalization as an int64. This is for backwards-compatibility with the old implementation.
  * I'm not actually sure if any connectors were producing this schema. All JDBC connectors were previously producing `{type: number}`, and now produce `{type: number, airbyte_type: integer}`.
* `{type: number, airbyte_type: integer}` is an int64
* `{type: string, airbyte_type: big_integer}` is an unbounded int
    * in normalization, we just use the `numeric` type, which hopefully is good enough. I'm not aware of DBs being able to represent actual unbounded ints.
    * The only source that actually produces big_integers is [source-zoho-crm](https://github.com/airbytehq/airbyte/blob/555efd53604b89e24d746080a546033e00c15329/airbyte-integrations/connectors/source-zoho-crm/source_zoho_crm/types.py#L149), so this isn't terribly relevant right now
    * Updated tests to use a 28-digit number, which is compatible with all destinations. BigQuery's NUMERIC type only supports up to 9E28.

### Testing
Normalization now verifies actual data values with some type tests. We probably should extend those to include timestamps, etc. But for the sake of getting this issue done, I only implemented tests for ints.

I ran some manual tests to verify that int/long columns are handled correctly:
* Postgres standard sync -> postgres
* Postgres CDC -> postgres
* mysql -> snowflake

## Recommended reading order
1. JsonSchemaType
2. `*SourceOperations` (JdbcSourceOperations, BigQuerySourceOperations, MysqlSourceOperations, PostgresSourceOperations, TiDbSourceOperations), plus tests:
    1. TestJdbcUtils
    2. CdcSourceTest
    3. Type tests (AbstractSourceDatabaseTypeTest, BigQuerySourceDatatypeTest, AbstractMysqlSourceDatatypeTest, CdcPostgresSourceDatatypeTest, PostgresSourceDatatypeTest)
    4. Acceptance tests (JdbcSourceAcceptanceTest, AbstractSshPostgresSourceAcceptanceTest, CdcPostgresSourceAcceptanceTest, PostgresJdbcSourceAcceptanceTest)
3. Normalization updates:
    1. datatypes.sql (afaik, only MySQL uses `signed` for the `int` type)
    2. utils.py (updated methods to separate int vs longs; most of this is boilerplate to pass around a full definition `{"type": "foo", "airbyte_type": "bar"}` rather than just the type value `"foo"`)
    3. stream_processor.py (mostly boilerplate to match the updated method signatures; note the new `is_long` call)
    4. everything in `base-normalization/integration_tests/resources`: This adds a new style of test, which actually verifies the output data values. Previously our tests only verified the number of rows written.
        1. The most interesting file is probably types_testing_incorrect_values.sql, which pulls the normalized data and compares it against the expected values
        2. catalog.json and messages.txt contain the input data to normalization.
4. destination-s3 - this is just bringing the S3 avro logic into compliance with our documentation
7. Test infra updates: (these were necessary to run them on my laptop)
    1. CdcPostgresSourceAcceptanceTest
    3. CdcPostgresSourceDatatypeTest
    6. PostgresSourceDatatypeTest
    8. PostgresSourceAcceptanceTest
8. `airbyte-integrations/bases/standard-destination-test/src/main/resources/*`, supported-data-types.md - reducing our big_integer tests from gigantic numbers to 28-digit numbers
9. normalization_test_output - most of this is irrelevant (a lot of it is actually cruft from older PRs); look at `types_testing_ab2.sql` and `types_testing_stg.sql` for the new type handling (specifically, the `as bigint` casts - TODO commit latest version where everything is `bigint`)

## 🚨 User Impact 🚨
None. Existing catalogs will be unchanged until users refresh their schema (at which point they'll be prompted to reset their data). Existing types behavior is (effectively) unchanged.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

### CODE REVIEW

File: airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/avro/JsonSchemaType.java [hunk 1]
```diff
@@ -16,8 +16,8 @@
 public enum JsonSchemaType {
 
   STRING("string", true, null, Schema.Type.STRING),
-  NUMBER_INT("number", true, "integer", Schema.Type.INT),
-  NUMBER_LONG("number", true, "big_integer", Schema.Type.LONG),
+  NUMBER_INT("number", true, "integer", Schema.Type.LONG),
+  NUMBER_LONG("number", true, "long", Schema.Type.STRING),
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/avro/JsonSchemaType.java:20]: @sashaNeshcheret are there already sources which produce `big_integer` as meaning an int64?

there's a project in-planning to change how we handle these types in general, so I'm not _too_ worried about long-term implications. Just want to make sure we don't break anything in the short/mid term.

Comment (edgao) [reply airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/avro/JsonSchemaType.java:20]: I'd like to reserve big_integer for an unbounded int (i.e. something similar to java's BigInteger class). Eventually I think we'll have something like this:
* `{type: ???, airbyte_type: int}` -> avro int
* `{type: ???, airbyte_type: long}` -> avro long
* `{type: ???, airbyte_type: big_integer}` -> avro string (? not sure if avro has an actual unbounded int type)

but we're still figuring out the details.

Comment (sashaNeshcheret) [reply airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/avro/JsonSchemaType.java:20]: Got your point regarding unbounded int, but such approach has disadvantages that we will lose type for all sources that really has range as big_integer and instead of using Schema.Type.LONG we will use Schema.Type.String. I believe we can add additional enum value for such case - NUMBER_BIG_INTEGER("number", true, "big_integer", Schema.Type.STRING), and map sources data types with range more than Long value to NUMBER_BIG_INTEGER and other to NUMBER_LONG, wdyt?

File: airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/avro/JsonSchemaType.java [hunk 2]
```diff
@@ -16,8 +16,9 @@
 public enum JsonSchemaType {
 
   STRING("string", true, null, Schema.Type.STRING),
-  NUMBER_INT("number", true, "integer", Schema.Type.INT),
-  NUMBER_LONG("number", true, "big_integer", Schema.Type.LONG),
+  // We currently don't distinguish between int and long
+  NUMBER_INT("number", true, "integer", Schema.Type.LONG),
+  NUMBER_BIGINT("number", true, "big_integer", Schema.Type.STRING),
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/avro/JsonSchemaType.java:21]: no sources actually produce big_integer yet, so we can freely change it from meaning int64 to meaning unbounded in t.

Comment (edgao) [reply airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/avro/JsonSchemaType.java:21]: actually, it looks like source-zoho-crm does (https://github.com/airbytehq/airbyte/blob/edgao/int_type/airbyte-integrations/connectors/source-zoho-crm/source_zoho_crm/types.py#L149) so that needs to remain supported.

**Comment by edgao** [comment]
mislicked. Not yet ready for review, sorry for early ping.

File: airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/avro/JsonSchemaType.java [hunk 3]
```diff
@@ -17,7 +17,7 @@ public enum JsonSchemaType {
 
   STRING("string", true, null, Schema.Type.STRING),
   NUMBER_INT("number", true, "integer", Schema.Type.INT),
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/avro/JsonSchemaType.java:19]: I think this (and the `INTEGER` entry below) should be switched to LONG, but had a _ton_ of trouble actually making that change. Given that current behavior is both wrong and unused, I feel fine with pushing this off until we finish the types revamp project. Given the uncertainty around what the final state looks like, will hold off on writing an issue for now.

---

for the curious: The problem is that we have some tests which test that we can handle unions of timestamp and integer. This resolves to an Avro union with types `["long", {"type": "long", "logicalType": "timestamp-micros"}]`. Unfortunately, Avro doesn't support this union, and complains that `long` is a duplicate entry.

But we also can't just drop one of those entries, because Avro expects a numeric input for `long`, but string input for the timestamp-micros. In other words:
* `{"created_at": 1634982000}` can be parsed as `"long"`, but not as `{"type": "long", "logicalType": "timestamp-micros"}`
* `{"created_at": "1634982000"}` (note the quotes around `1634982000`) can be parsed as `{"type": "long", "logicalType": "timestamp-micros"}`, but not as `"long"`.

File: docs/understanding-airbyte/supported-data-types.md [hunk 4]
```diff
@@ -17,14 +17,16 @@ This table summarizes the available types. See the [Specific Types](#specific-ty
...
 | Array                                                          | `{"type": "array"}`; optionally `items` and `additionalItems`                            | `[1, 2, 3]`                                                                     |
 | Object                                                         | `{"type": "object"}`; optionally `properties` and `additionalProperties`                 | `{"foo": "bar"}`                                                                |
 | Untyped (i.e. any value is valid)                              | `{}`                                                                                     |                                                                                 |
 | Union                                                          | `{"anyOf": [...]}` or `{"oneOf": [...]}`                                                 |                                                                                 |
 
+Note that some of these may be destination-dependent. For example, Snowflake `NUMERIC` columns can be at most 38 digits wide, but Postgres `NUMERIC` columns may have up to 131072 digits before the decimal point.
```
Comment (edgao) [inline docs/understanding-airbyte/supported-data-types.md:28]: this is something that needs to be addressed as part of the types project. Kind of similar to https://github.com/airbytehq/airbyte/issues/14756, expect that numbers are typically better at having easily-defined ranges, so we can probably handle these more smartly.

**Comment by edgao** [comment]
@airbytehq/jdbc-connectors / @airbytehq/normalization : current test failures are entirely coming from the big_integer handling:
* destination-snowflake - the destination is actually working correctly. The failure is because when the test retrieves the data from the destination, the snowflake-jdbc driver incorrectly interprets the `NUMBER` column as an int64. See https://github.com/snowflakedb/snowflake-jdbc/issues/252
* base-normalization:
  * mssql - `('22003', '[22003] [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Arithmetic overflow error converting nvarchar to data type numeric. (8115) (SQLMoreResults)')`
  * clickhouse - populating `null` as the big_integer values for unknown reason
  * mysql - passes locally after https://github.com/airbytehq/airbyte/pull/14362/commits/599bb18df86cc93814af4948e52060c96f659ba8#diff-d62e4a8a19e6547243431d0d09aea84f7653ab6fe1947236e0d822cac1334eeaR141-R143

I'll spend some time next week trying to get these working, but they're pretty low-priority given that this functionality is not currently in use. If it takes too long I might just wipe out the big_integer tests entirely (especially since we're going to rework the type system this quarter).

**Comment by grishick** [comment]
> @airbytehq/jdbc-connectors / @airbytehq/normalization : current test failures are entirely coming from the big_integer handling:
> 
> * destination-snowflake - the destination is actually working correctly. The failure is because when the test retrieves the data from the destination, the snowflake-jdbc driver incorrectly interprets the `NUMBER` column as an int64. See [SNOW-165204: BigDecimal specified as java.sql.Types.BIGINT snowflakedb/snowflake-jdbc#252](https://github.com/snowflakedb/snowflake-jdbc/issues/252)
> * base-normalization:
>   
>   * mssql - `('22003', '[22003] [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Arithmetic overflow error converting nvarchar to data type numeric. (8115) (SQLMoreResults)')`
>   * clickhouse - populating `null` as the big_integer values for unknown reason
>   * mysql - passes locally after [599bb18#diff-d62e4a8a19e6547243431d0d09aea84f7653ab6fe1947236e0d822cac1334eeaR141-R143](https://github.com/airbytehq/airbyte/commit/599bb18df86cc93814af4948e52060c96f659ba8#diff-d62e4a8a19e6547243431d0d09aea84f7653ab6fe1947236e0d822cac1334eeaR141-R143)
> 
> I'll spend some time next week trying to get these working, but they're pretty low-priority given that this functionality is not currently in use. If it takes too long I might just wipe out the big_integer tests entirely (especially since we're going to rework the type system this quarter).

Could you disable the broken tests until they are fixed, if you think they are low priority? Otherwise, they'll keep slowing down PRs.

**Comment by edgao** [comment]
> disable the broken tests until they are fixed

yep, if I don't get them working today I'll disable them as part of this PR.

**Comment by edgao** [comment]
Will make a note to revert https://github.com/airbytehq/airbyte/pull/14362/commits/f03935c6b9b5616236c0aa3aaa19536afed3bcb8 as part of the types project.

File: airbyte-integrations/bases/base-normalization/integration_tests/normalization_test_output/redshift/test_nested_streams/first_output/airbyte_incremental/test_normalization/nested_stream_with_complex_columns_resulting_into_long_names_partition.sql [hunk 5]
```diff
@@ -2,7 +2,7 @@
       
 
   create  table
-    "integrationtests"."test_normalization"."nested_stream_with_complex_columns_resulting_into_long_names_partition__dbt_tmp"
+    "integrationtests".test_normalization_lpgvp."nested_stream_with_complex_columns_resulting_into_long_names_partition"
```
Comment (grishick) [inline airbyte-integrations/bases/base-normalization/integration_tests/normalization_test_output/redshift/test_nested_streams/first_output/airbyte_incremental/test_normalization/nested_stream_with_complex_columns_resulting_into_long_names_partition.sql:5]: is this change relevant to handling ints and longs?

**Review by grishick — APPROVED** [review]

**Comment by edgao** [comment]
/publish connector=bases/base-normalization,connectors/source-bigquery,connectors/source-postgres,connectors/source-mysql,connectors/source-tidb parallel=true

File: airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/avro/JsonSchemaType.java (hunk 3 above)
Comment (rshanmugam1) [reply airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/avro/JsonSchemaType.java:19]: @edgao it was having problem with bigint. 

postgres type int8 (bigint)
airbyte schema  ended up- {"type": "number", "airbyte_type": "integer"}

in parquet ended up integer and long values trimmed down and data is unusable. i had to patch to Long to get it work.
